src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,160 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private static final String EXCEPTION_MESSAGE = "The index is out of bounds";
+    private Node<T> first;
+    private Node<T> last;
+    private int size;
+
     @Override
     public boolean add(T value) {
-        return false;
+        Node<T> newNode = new Node<>(value);
+        if (size == 0) {
+            first = newNode;
+            last = newNode;
+        } else {
+            newNode.prev = last;
+            last.next = newNode;
+            last = newNode;
+        }
+        size++;
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException(EXCEPTION_MESSAGE);
+        }
+        if (index == size) {
+            add(value);
+        } else {
+            Node<T> currentNode = searchNodeByIndex(index);
+            Node<T> previousNode = currentNode.prev;
+            Node<T> newNode = new Node<>(value);
+            newNode.prev = previousNode;
+            newNode.next = currentNode;
+            currentNode.prev = newNode;
+            if (previousNode == null) {
+                first = newNode;
+            } else {
+                previousNode.next = newNode;
+            }
+            size++;
+        }
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        if (list == null || list.isEmpty()) {
+            return false;
+        }
+        for (T listItem : list) {
+            add(listItem);
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (first == null) {
+            throw new IndexOutOfBoundsException(EXCEPTION_MESSAGE);
+        }
+        return searchNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> currentNode = searchNodeByIndex(index);
+        T oldValue = currentNode.value;
+        currentNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
```

Comment:
you already do this in `searchNodeByIndex`

Reply:
done